src/bin/main.rs

```diff
@@ -1,12 +1,117 @@
 
-fn main() {
-    match binaries("Cargo.toml", Build::Debug) {
-        Ok(binaries) => {
-            for binary in binaries {
-                println!("{}", binary.to_str().unwrap());
+/// Build the command line interface.
+///
+/// The CLI currently supports the distinction between debug and release builds
+/// (selected via the `--release` flag) as well as the selection of the target
+/// to execute. Currently binaries, examples and benches are supported.
+fn cli<'a, 'b>() -> App<'a, 'b> {
+    App::new(crate_name!())
+        .about("Cargo subcommand for running valgrind")
+        .author(crate_authors!())
+        .version(crate_version!())
+        .arg(
+            Arg::with_name("release")
+                .help("Build and run artifacts in release mode, with optimizations")
+                .long("release"),
+        )
+        .arg(
+            Arg::with_name("bin")
+                .help("Build and run the specified binary")
+                .long("bin")
+                .takes_value(true)
+                .value_name("NAME")
+                .conflicts_with_all(&["example", "bench"]),
+        )
+        .arg(
+            Arg::with_name("example")
+                .help("Build and run the specified example")
+                .long("example")
+                .takes_value(true)
+                .value_name("NAME")
+                .conflicts_with_all(&["bin", "bench"]),
+        )
+        .arg(
+            Arg::with_name("bench")
+                .help("Build and run the specified bench")
+                .long("bench")
+                .takes_value(true)
+                .value_name("NAME")
+                .conflicts_with_all(&["bin", "example"]),
+        )
+        .arg(
+            Arg::with_name("manifest")
+                .help("Path to Cargo.toml")
+                .long("manifest-path")
+                .takes_value(true)
+                .value_name("PATH"),
+        )
+}
+
+fn run() -> Result<bool, Box<dyn std::error::Error>> {
+    let cli = cli().get_matches();
+    let build = if cli.is_present("release") {
+        Build::Release
+    } else {
+        Build::Debug
+    };
+    let binary = cli
+        .value_of("bin")
+        .or(cli.value_of("example"))
+        .or(cli.value_of("bench"));
+    let manifest = cli.value_of("manifest").unwrap_or("Cargo.toml".into());
+    let manifest = PathBuf::from(manifest).canonicalize()?;
+
+    let binaries = binaries(&manifest, build)?;
+    let binary = match binary {
+        Some(path) => PathBuf::from(path),
+        None if binaries.len() == 1 => PathBuf::from(&binaries[0]),
+        None => Err("Multiple possible targets, please specify more precise")?,
+    };
+    let binary = binaries
+        .into_iter()
+        .find(|path| path.file_name() == binary.file_name())
+        .ok_or("Could not find selected binary")?;
+    let target = binary.file_name().unwrap().into();
+    let target = Target::Binary(target); // FIXME: use correct variant
```

Comment:
Note, that currently only binary targets are supported due to this unconditional line.